### PR TITLE
Use `tempfile` in `mlflow.log_artifact` example

### LINKED ISSUE #11093: [FR] Add functionality to log artifact from memory
### Willingness to contribute

Yes. I can contribute this feature with guidance from the MLflow community.

### Proposal Summary

Give users the ability to log artifacts from memory.

Example
```python
# pip install mlflow kaleido -U -q
import mlflow
import plotly.graph_objects as go

with mlflow.start_run():
  fig = go.Figure(data=[go.Bar(y=[2, 3, 1])])

  with tempfile.NamedTemporaryFile(suffix=".png") as tmpfile:
    fig.write_image(tmpfile.name)
    mlflow.log_artifact(tmpfile.name, "xyz.png")
```

The core challenge is supporting object-specific serialization methods. Patching the write method for every python object isn't scalable and passing the object's write method to a `log_artifact_from_memory` function is a bit "low level" for data scientists. I'd appreciate any guidance on whether this is possible. 

If it's not possible, we should at least modify the [fluent log_artfact](https://github.com/mlflow/mlflow/blob/dc51fc6b82fa78a30f8be5efed18f7867809311d/mlflow/tracking/fluent.py#L1010) docstring to use a tempfile and add some notes in the tutorials. Duplicating data is bad.

### Motivation

> #### What is the use case for this feature?
I want to log a plotly figure. The below solutions are unnecessarily complex and should be abstracted from the user
1. Use a tempfile with `mlflow.log_artifact` 
2. Convert to an object format that supports serialization with log_image/text/dict

> #### Why is this use case valuable to support for MLflow users in general?
Doing the above workarounds introduce friction into the product.

> #### Why is this use case valuable to support for your project(s) or organization?

> #### Why is it currently difficult to achieve this use case?

### Details

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/11095?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11095/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11095
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Resolve #11093

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thanks Haru!

File: mlflow/tracking/fluent.py [hunk 1]
```diff
@@ -1053,21 +1057,24 @@ def log_artifacts(
...
-        import os
+        import tempfile
+        from pathlib import Path
         import mlflow
 
         # Create some files to preserve as artifacts
         features = "rooms, zipcode, median_price, school_rating, transport"
         data = {"state": "TX", "Available": 25, "Type": "Detached"}
         # Create couple of artifact files under the directory "data"
         os.makedirs("data", exist_ok=True)
```
Comment (BenWilson2) [inline mlflow/tracking/fluent.py:1068]: I think we need to delete this line :)

Comment (BenWilson2) [reply mlflow/tracking/fluent.py:1068]:
```suggestion
```

Comment (harupy) [reply mlflow/tracking/fluent.py]: thanks for the catch!

**Pull request merged**